### PR TITLE
Node 12 (JITless) support

### PR DESCRIPTION
**Pull request by modulesio**
V8 now supports [running in JITless mode](https://v8.dev/blog/jitless) and there is a PR to [bring it into node](https://github.com/nodejs/node/pull/26685).

With this feature, we can run Exokit on more platforms, including iOS, which disallow the executable writable memory policy that V8 uses by default.

The main task for us in exoland is to make sure we are compatible with Node 12. The V8 API changed a bit re: `v8::Handle` here, and some deprecated methods were removed. This PR makes the necessary changes to exokit and collects the needed external module updates.

Notably we would need to do this work anyway when Node 12 comes around in April.

Continuation of https://github.com/webmixedreality/exokit/pull/871; thanks to @shawwn for the research and bootstrapping that work.

### CODE REVIEW

**Comment by modulesio** [comment]
This should be building on node 12 now.

File: deps/exokit-bindings/webaudiocontext/src/AudioContext.cpp [hunk 1]
```diff
@@ -427,14 +427,14 @@ NAN_METHOD(AudioContext::CreateBufferSource) {
 NAN_METHOD(AudioContext::CreateScriptProcessor) {
   Nan::HandleScope scope;
 
-  uint32_t bufferSize = info[0]->IsNumber() ? info[0]->Uint32Value() : 256;
-  uint32_t numberOfInputChannels = info[1]->IsNumber() ? info[1]->Uint32Value() : 2;
-  uint32_t numberOfOutputChannels = info[2]->IsNumber() ? info[2]->Uint32Value() : 2;
+  uint32_t bufferSize = info[0]->IsNumber() ? TO_UINT32(info[0]) : 256;
```
Comment (shawwn) [inline deps/exokit-bindings/webaudiocontext/src/AudioContext.cpp]: Should `info.Length() >= 3` be checked here?

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 2]
```diff
@@ -1820,12 +1820,13 @@ NAN_METHOD(WebGLRenderingContext::Uniform3uiv) {
       data = getArrayData<GLuint>(dataValue, &count);
     }
     if (info[2]->IsNumber()) {
-      GLsizei srcOffset = info[3]->Uint32Value();
+
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: extra whitespace

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 3]
```diff
@@ -2519,8 +2520,8 @@ NAN_METHOD(WebGLRenderingContext::GetProgramParameter) {
 
 
 NAN_METHOD(WebGLRenderingContext::GetUniformLocation) {
-  GLint programId = info[0]->ToObject()->Get(JS_STR("id"))->Int32Value();
-  v8::String::Utf8Value name(info[1]);
+  GLint programId = TO_INT32(JS_OBJ(info[0])->Get(JS_STR("id")));
+  Nan::Utf8String name(info[1]);
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: inconsistent with the rest of the Nan::Utf8String lines (should `info[1]` be casted to a string here?)

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 4]
```diff
@@ -3079,27 +3080,27 @@ NAN_METHOD(WebGLRenderingContext::CompressedTexImage2D) {
 }
 
 NAN_METHOD(WebGLRenderingContext::TexParameteri) {
-  int target = info[0]->Int32Value();
-  int pname = info[1]->Int32Value();
-  int param = info[2]->Int32Value();
+  int target = TO_INT32(info[0]);
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: should `info.Length()` be checked here?

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 5]
```diff
@@ -3446,41 +3447,41 @@ NAN_METHOD(WebGLRenderingContext::Finish) {
...
+  GLfloat x = TO_FLOAT(info[1]);
 
   glVertexAttrib1f(indx, x);
 }
 
 NAN_METHOD(WebGLRenderingContext::VertexAttrib2f) {
-  GLuint indx = info[0]->Int32Value();
-  float x = (float)info[1]->NumberValue();
-  float y = (float)info[2]->NumberValue();
+  GLuint indx = TO_UINT32(info[0]);
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: Should this be `TO_INT32` or was this intentional to match `GLuint` type?

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 6]
```diff
@@ -3446,41 +3447,41 @@ NAN_METHOD(WebGLRenderingContext::Finish) {
...
 
   glVertexAttrib2f(indx, x, y);
 }
 
 NAN_METHOD(WebGLRenderingContext::VertexAttrib3f) {
-  GLuint indx = info[0]->Int32Value();
-  float x = (float)info[1]->NumberValue();
-  float y = (float)info[2]->NumberValue();
-  float z = (float)info[3]->NumberValue();
+  GLuint indx = TO_INT32(info[0]);
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: Should this be `TO_UINT32` to be consistent with line 3457?

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 7]
```diff
@@ -3985,8 +3986,8 @@ NAN_METHOD(WebGLRenderingContext::GetVertexAttribOffset) {
 }
 
 NAN_METHOD(WebGLRenderingContext::GetShaderPrecisionFormat) {
-  GLenum shaderType = info[0]->Uint32Value();
-  GLenum precisionType = info[1]->Uint32Value();
+  GLenum shaderType = TO_UINT32(info[0]);
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: should these be `TO_INT32`? (elsewhere you use `TO_INT32` for casting to `GLenum`)

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 8]
```diff
@@ -4124,27 +4125,27 @@ NAN_METHOD(WebGLRenderingContext::GetShaderSource) {
...
   WebGLRenderingContext *gl = ObjectWrap::Unwrap<WebGLRenderingContext>(info.This());
-  GLenum targetV = info[0]->Uint32Value();
-  GLint levelV = info[1]->Int32Value();
-  GLint xoffsetV = info[2]->Int32Value();
-  GLint yoffsetV = info[3]->Int32Value();
-  GLsizei widthV = info[4]->Uint32Value();
-  GLsizei heightV = info[5]->Uint32Value();
-  GLenum formatV = info[6]->Uint32Value();
-  GLenum typeV = info[7]->Uint32Value();
+  GLenum targetV = TO_UINT32(info[0]);
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: should this be `TO_INT32` for casting to `GLenum`?

File: deps/exokit-bindings/webglcontext/src/webgl.cc (hunk 7 above)
Comment (modulesio) [reply deps/exokit-bindings/webglcontext/src/webgl.cc]: Yes, thank you.

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 9]
```diff
@@ -4896,7 +4897,7 @@ NAN_METHOD(WebGLRenderingContext::GetSupportedExtensions) {
 
 // TODO GetExtension(name) return the extension name if found, should be an object...
 NAN_METHOD(WebGLRenderingContext::GetExtension) {
-  String::Utf8Value name(info[0]);
+  Nan::Utf8String name(info[0]);
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: should `info[0]` be `::Cast`'ed to a string here?

File: deps/exokit-bindings/webglcontext/src/webgl.cc (hunk 8 above)
Comment (modulesio) [reply deps/exokit-bindings/webglcontext/src/webgl.cc]: Yup.

File: deps/exokit-bindings/webaudiocontext/src/AudioContext.cpp (hunk 1 above)
Comment (modulesio) [reply deps/exokit-bindings/webaudiocontext/src/AudioContext.cpp]: I think the number extraction is sufficient for now. Guarding all cases against invalid arguments would be another holistic effort in these APIs.

File: deps/exokit-bindings/webglcontext/src/webgl.cc (hunk 5 above)
Comment (modulesio) [reply deps/exokit-bindings/webglcontext/src/webgl.cc]: I think the spec says `GLuint` for all of these: https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/vertexAttrib#Parameters

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 10]
```diff
@@ -5401,31 +5402,31 @@ NAN_METHOD(WebGL2RenderingContext::IsSampler) {
...
 
   glBindSampler(unit, sampler);
 }
 
 NAN_METHOD(WebGL2RenderingContext::SamplerParameteri) {
-  GLuint sampler = info[0]->ToObject()->Get(JS_STR("id"))->Uint32Value();
-  GLenum pname = info[1]->Int32Value();
-  GLint param =  info[2]->Int32Value();
+  GLuint sampler = TO_UINT32(JS_OBJ(info[0])->Get(JS_STR("id")));
+  GLenum pname = TO_UINT32(info[1]);
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: Should this be `TO_INT32`?

File: deps/exokit-bindings/webglcontext/src/webgl.cc [hunk 11]
```diff
@@ -5401,31 +5402,31 @@ NAN_METHOD(WebGL2RenderingContext::IsSampler) {
...
+  GLfloat param = TO_FLOAT(info[2]);
     
   glSamplerParameterf(sampler, pname, param);
 }
 
 NAN_METHOD(WebGL2RenderingContext::GetSamplerParameter) {
-  GLuint sampler = info[0]->ToObject()->Get(JS_STR("id"))->Uint32Value();
-  GLenum pname = info[1]->Int32Value();
+  GLuint sampler = TO_UINT32(JS_OBJ(info[0])->Get(JS_STR("id")));
+  GLenum pname = TO_UINT32(info[1]);
```
Comment (shawwn) [inline deps/exokit-bindings/webglcontext/src/webgl.cc]: Should this be `TO_INT32`?

File: deps/exokit-bindings/windowsystem/src/windowsystem.cc [hunk 12]
```diff
@@ -417,12 +417,12 @@ bool CreateRenderTarget(WebGLRenderingContext *gl, int width, int height, GLuint
 
 NAN_METHOD(CreateRenderTarget) {
   WebGLRenderingContext *gl = ObjectWrap::Unwrap<WebGLRenderingContext>(Local<Object>::Cast(info[0]));
-  int width = info[1]->Uint32Value();
-  int height = info[2]->Uint32Value();
-  GLuint sharedColorTex = info[3]->Uint32Value();
-  GLuint sharedDepthStencilTex = info[4]->Uint32Value();
-  GLuint sharedMsColorTex = info[5]->Uint32Value();
-  GLuint sharedMsDepthStencilTex = info[6]->Uint32Value();
+  int width = TO_UINT32(info[1]);
```
Comment (shawwn) [inline deps/exokit-bindings/windowsystem/src/windowsystem.cc]: signed/unsigned mismatch

File: deps/exokit-bindings/windowsystem/src/windowsystem.cc [hunk 13]
```diff
@@ -450,14 +450,14 @@ NAN_METHOD(CreateRenderTarget) {
...
   WebGLRenderingContext *gl = ObjectWrap::Unwrap<WebGLRenderingContext>(Local<Object>::Cast(info[0]));
-  int width = info[1]->Uint32Value();
-  int height = info[2]->Uint32Value();
-  GLuint fbo = info[3]->Uint32Value();
-  GLuint colorTex = info[4]->Uint32Value();
-  GLuint depthStencilTex = info[5]->Uint32Value();
-  GLuint msFbo = info[6]->Uint32Value();
-  GLuint msColorTex = info[7]->Uint32Value();
-  GLuint msDepthStencilTex = info[8]->Uint32Value();
+  int width = TO_UINT32(info[1]);
```
Comment (shawwn) [inline deps/exokit-bindings/windowsystem/src/windowsystem.cc]: signed/unsigned mismatch

File: exokit.cpp [hunk 14]
```diff
@@ -32,11 +32,11 @@ void callFunction(const char *funcname, const int argc, Local<Value> argv[]) {
 
   // call function, 'this' points to global object
   TryCatch try_catch(Isolate::GetCurrent());
-  Local<Value> result = jsfunc->Call(global, argc, argv);
+  Local<Value> result = Nan::Call(jsfunc, global, argc, argv).ToLocalChecked();
```
Comment (shawwn) [inline exokit.cpp]: Is it intentional to use `Nan::Call` here rather than doing `jsfunc->Call(JS_CONTEXT(), ...)`?

Comment (modulesio) [reply exokit.cpp]: I think it should be the same thing, but it would be good to be consistent.

**Comment by modulesio** [comment]
Note that to use this feature, you probably want the Node 12 `--jitless` flag:

```
  --jitless (Disable runtime allocation of executable memory.)
        type: bool  default: false
```

**Pull request merged**